Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/image/ImageLoaderV2.java

```
@@ -0,0 +1,164 @@
+import android.graphics.BitmapFactory;
+import android.os.Handler;
+import android.os.Looper;
+
+import com.android.volley.VolleyError;
+import com.android.volley.toolbox.ImageLoader;
+import com.github.adamantcheese.chan.core.manager.ThreadSaveManager;
+import com.github.adamantcheese.chan.core.model.PostImage;
+import com.github.adamantcheese.chan.core.model.orm.Loadable;
+import com.github.adamantcheese.chan.core.settings.ChanSettings;
+import com.github.adamantcheese.chan.utils.Logger;
+import com.github.adamantcheese.chan.utils.StringUtils;
+
+import java.io.File;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class ImageLoaderV2 {
+    private static final String TAG = "ImageLoaderV2";
+
+    private ImageLoader imageLoader;
+    private Executor diskLoaderExecutor = Executors.newSingleThreadExecutor();
+    private Handler mainThreadHandler = new Handler(Looper.getMainLooper());
+
+    public ImageLoaderV2(ImageLoader imageLoader) {
+        this.imageLoader = imageLoader;
+    }
+
+    public ImageContainer getImage(
+            boolean isThumbnail,
+            Loadable loadable,
+            PostImage postImage,
+            int width,
+            int height,
+            ImageListener imageListener) {
+        if (loadable.isSavedCopy) {
+            String formattedName;
+
+            if (postImage.spoiler) {
+                String extension = StringUtils.extractFileExtensionFromImageUrl(
+                        postImage.spoilerThumbnailUrl.toString());
+
+                formattedName = ThreadSaveManager.formatSpoilerImageName(extension);
+            } else {
+                if (isThumbnail) {
+                    String extension = StringUtils.extractFileExtensionFromImageUrl(
+                            postImage.thumbnailUrl.toString());
+
+                    if (extension == null) {
+                        Logger.e(TAG, "Could not extract extension from a thumbnailUrl = "
```

Probably should throw an exception here.